client/network/src/service/out_events.rs

```diff
@@ -0,0 +1,276 @@
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Event>> {
+		if let Some(ev) = ready!(Pin::new(&mut self.inner).poll_next(cx)) {
+			let metrics = self.metrics.lock().clone();
+			if let Some(Some(metrics)) = metrics.as_ref().map(|m| &**m) {
+				metrics.event_out(&ev);
+			} else {
+				log::warn!("Inconsistency in out_events: event happened before sender associated");
+			}
+			Poll::Ready(Some(ev))
+		} else {
+			Poll::Ready(None)
+		}
+	}
+}
+
+impl fmt::Debug for Receiver {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Receiver").finish()
+	}
+}
+
+impl Drop for Receiver {
+	fn drop(&mut self) {
+		// Empty the list to properly decrease the metrics.
+		while let Some(Some(_)) = self.next().now_or_never() {}
+	}
+}
+
+/// Collection of senders.
+pub struct OutChannels {
+	event_streams: Vec<Sender>,
+	/// The metrics we collect. A clone of this is sent to each [`Receiver`] associated with this
+	/// object.
+	metrics: Arc<Option<Metrics>>,
+}
+
+impl OutChannels {
+	/// Creates a new empty collection of senders.
+	pub fn new(registry: Option<&Registry>) -> Result<Self, PrometheusError> {
+		let metrics = if let Some(registry) = registry {
+			Some(Metrics::register(registry)?)
+		} else {
+			None
+		};
+
+		Ok(OutChannels {
+			event_streams: Vec::new(),
+			metrics: Arc::new(metrics),
+		})
+	}
+
+	/// Adds a new [`Sender`] to the collection.
+	pub fn push(&mut self, sender: Sender) {
+		let mut metrics = sender.metrics.lock();
+		debug_assert!(metrics.is_none());
+		*metrics = Some(self.metrics.clone());
+		drop(metrics);
+		self.event_streams.push(sender);
+
+		if let Some(metrics) = &*self.metrics {
+			metrics.out_events_num_channels.inc();
+		}
+	}
+
+	/// Sends an event.
+	pub fn send(&mut self, event: Event) {
+		self.event_streams.retain(|sender| {
+			sender.inner.unbounded_send(event.clone()).is_ok()
+		});
+
+		// The number of elements remaining in `event_streams` corresponds to the number of times
+		// we have sent an element on the channels.
+		if let Some(metrics) = &*self.metrics {
+			metrics.event_in(&event, self.event_streams.len() as u64);
+		}
+	}
+}
+
+impl fmt::Debug for OutChannels {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("OutChannels")
+			.field("num_channels", &self.event_streams.len())
+			.finish()
+	}
+}
+
+struct Metrics {
+	// This list is ordered alphabetically
+	out_events_num_channels: Gauge<U64>,
+	out_events_in: CounterVec<U64>,
+	out_events_notifications_sizes: GaugeVec<U64>,
+	out_events_out: CounterVec<U64>,
+}
+
+impl Metrics {
+	fn register(registry: &Registry) -> Result<Self, PrometheusError> {
+		Ok(Self {
+			out_events_num_channels: register(Gauge::new(
+				"sub_libp2p_out_events_num_channels",
+				"Number of internal active channels that broadcast network events",
+			)?, registry)?,
+			out_events_in: register(CounterVec::new(
+				Opts::new(
+					"sub_libp2p_out_events_notifications_opened_count",
```

Comment:
I think this metric name is outdated, right? Shouldnt it be something like 
```suggestion
					"sub_libp2p_out_events_notifications_count",
```